btree/btree.c

```diff
@@ -3529,164 +3565,561 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+			node_op_fini(&oi->i_nop);
+			return fail(bop, M0_ERR(-EFAULT));
+		}
+
+		return node_get(&oi->i_nop, bop->bo_arbor->t_desc,
+				&root_child, P_STORE_CHILD);
+	}
+	return P_LOCK;
 }
 
-#endif
+/* State machine implementation for delete operation */
+static int64_t btree_del_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop        = M0_AMB(bop, smop, bo_op);
+	struct td             *tree       = bop->bo_arbor->t_desc;
+	uint64_t               flags      = bop->bo_flags;
+	struct m0_btree_oimpl *oi         = bop->bo_i;
+	struct level          *lev;
 
-#ifndef __KERNEL__
-/**
- *  --------------------------
- *  Section START - Unit Tests
- *  --------------------------
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((flags & BOF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie) &&
+		    !node_isunderflow(oi->i_cookie_node, true))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	case P_SETUP: {
+		bop->bo_arbor->t_height = tree->t_height;
+		bop->bo_i = level_alloc(bop->bo_arbor->t_height);
+		if (bop->bo_i == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		bop->bo_i->i_key_found = false;
+		return P_LOCKALL;
+	}
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL) {
+			/* return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN); */
+			m0_rwlock_write_lock(&tree->t_lock);
+		}
+		/* Fall through to the next stage */
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    node_slot = {};
+			struct segaddr child_node_addr;
+
+			lev = &oi->i_level[oi->i_used];
+			lev->l_node = oi->i_nop.no_node;
+			node_slot.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			/* verify node footer */
+			int rc = node_verify(lev->l_node);
+			if (rc)
+			{
+				node_op_fini(&oi->i_nop);
+				return fail(bop, rc);
+			}
+			oi->i_nop.no_node = NULL;
+
+			oi->i_key_found = node_find(&node_slot,
+						    &bop->bo_rec.r_key);
+			lev->l_idx = node_slot.s_idx;
+
+			if (node_level(node_slot.s_node) > 0) {
+				if (oi->i_key_found) {
+					lev->l_idx++;
+					node_slot.s_idx++;
+				}
+				node_child(&node_slot, &child_node_addr);
+
+				if (!address_in_segment(child_node_addr)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child_node_addr, P_NEXTDOWN);
+			} else {
+				if (!oi->i_key_found)
+					return P_LOCK;
+				/**
+				 * If root is an internal node and it contains
+				 * only two record, if any of the record is
+				 * going to be deleted, load the other child of
+				 * root.
+				 */
+				if (oi->i_used > 0 &&
+				    node_count_rec(oi->i_level[0].l_node) == 2)
+					return root_case_handle(bop);
```

Comment:
Why is this special case needed here? We are already traversing the the levels (from leaf to root) in P_ACT(btree_del_resolve_underflow) and fixing underflows. Also we have checks for specificcally this case in P_ACT too. We can save the child node pointers there and avoid traversing twice.

Reply:
We can not directly load the child node of root, first we need to check if there will be underflow at root node or not, if yes, then only we are going to load the child node

Reply:
Yes, but don't we have the same checks in P_ACT once we reach root node. 
I just feel we are introducing extra checks here, It could have been done in P_ACT. Anyways if other's don't have any issue then I will let it go.

Reply:
Yes we have the same checks in P_ACT, in P_ACT(btree_del_resolve_underflow)  we start resolving underflow from bottom to root, but we can not load the child node in P_ACT as it will get executed in lock mode, So, we have to load the root child before lock mode, and before that we have to check will there be underflow at root, therefore root_case_handle is needed before lock

Reply:
Well it would have been only one node loading so not much of overhead. Lets keep it as it is for now and see if we can optimize it later.